{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On this lab we will put to practice some of the concepts we have learned on this past few days.\n",
    "\n",
    "`NOTE: On this lab you should try to write all the functions yourself using only the most basic of python syntax and without functions such as len, count, sum, max, min, in, etc. Give it a try. 🧑🏻‍💻👩🏻‍💻`\n",
    "\n",
    "The cell after each exercise contains a few tests to check if your function works as expected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from mod.testing import *\n",
    "import unittest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Write a function that returns the greater of two numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Find which number is greater\n",
      "Enter the first number: 6\n",
      "Enter the second number: 5\n",
      "5  is greater than  3\n"
     ]
    }
   ],
   "source": [
    "def greater(a,b):\n",
    "    if (a > b):\n",
    "        print(a,\" is greater than \",b)\n",
    "    else:\n",
    "        print(b,\" is greater than \",a)\n",
    "\n",
    "print(\"Find which number is greater\")\n",
    "a = input(\"Enter the first number: \")\n",
    "b = input(\"Enter the second number: \")\n",
    "greater(num1, num2)  \n",
    "\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 679 : Should be 679\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 824 : Should be 824\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 916 : Should be 916\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 892 : Should be 892\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 154 : Should be 154\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -572 : Should be -572\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 738 : Should be 738\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 555 : Should be 555\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 300 : Should be 300\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -897 : Should be -897\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 769 : Should be 769\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 662 : Should be 662\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 699 : Should be 699\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -38 : Should be -38\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -497 : Should be -497\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -240 : Should be -240\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 709 : Should be 709\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -112 : Should be -112\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -381 : Should be -381\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -462 : Should be -462\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 427 : Should be 427\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 424 : Should be 424\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 13 : Should be 13\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -218 : Should be -218\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 178 : Should be 178\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 378 : Should be 378\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 248 : Should be 248\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 792 : Should be 792\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 554 : Should be 554\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 792 : Should be 792\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 574 : Should be 574\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 771 : Should be 771\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 444 : Should be 444\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -518 : Should be -518\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -291 : Should be -291\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 253 : Should be 253\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 192 : Should be 192\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 943 : Should be 943\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 566 : Should be 566\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 731 : Should be 731\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 150 : Should be 150\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -311 : Should be -311\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -517 : Should be -517\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -418 : Should be -418\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 504 : Should be 504\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 929 : Should be 929\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 994 : Should be 994\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 121 : Should be 121\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 976 : Should be 976\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 709 : Should be 709\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 659 : Should be 659\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -664 : Should be -664\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 407 : Should be 407\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -374 : Should be -374\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 464 : Should be 464\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -382 : Should be -382\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 283 : Should be 283\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 510 : Should be 510\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 808 : Should be 808\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 731 : Should be 731\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 666 : Should be 666\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 358 : Should be 358\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 366 : Should be 366\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 946 : Should be 946\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 50 : Should be 50\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 224 : Should be 224\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -685 : Should be -685\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 605 : Should be 605\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 459 : Should be 459\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 453 : Should be 453\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 2 : Should be 2\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 494 : Should be 494\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -268 : Should be -268\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 37 : Should be 37\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 857 : Should be 857\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -121 : Should be -121\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 845 : Should be 845\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 509 : Should be 509\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 705 : Should be 705\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 744 : Should be 744\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 944 : Should be 944\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 564 : Should be 564\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 391 : Should be 391\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 980 : Should be 980\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -852 : Should be -852\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 884 : Should be 884\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 685 : Should be 685\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -256 : Should be -256\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 121 : Should be 121\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -337 : Should be -337\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -42 : Should be -42\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 890 : Should be 890\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -113 : Should be -113\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -460 : Should be -460\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 557 : Should be 557\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 636 : Should be 636\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -430 : Should be -430\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 590 : Should be 590\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != -498 : Should be -498\n",
      "\n",
      "======================================================================\n",
      "FAIL: runTest (mod.testing.test_greater.<locals>.TestKnown)\n",
      "----------------------------------------------------------------------\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/queraltiglesias/Desktop/Data/LAB/week_1/lab-functions/mod/testing.py\", line 17, in runTest\n",
      "    self.assertEqual(fn(*self.input), self.output, f\"Should be {self.output}\")\n",
      "AssertionError: None != 256 : Should be 256\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.075s\n",
      "\n",
      "FAILED (failures=100)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "679  is greater than  -947\n",
      "824  is greater than  782\n",
      "916  is greater than  -905\n",
      "892  is greater than  511\n",
      "154  is greater than  -749\n",
      "-572  is greater than  -870\n",
      "738  is greater than  585\n",
      "555  is greater than  -35\n",
      "300  is greater than  117\n",
      "-897  is greater than  -929\n",
      "769  is greater than  -631\n",
      "662  is greater than  337\n",
      "699  is greater than  -592\n",
      "-38  is greater than  -440\n",
      "-497  is greater than  -834\n",
      "-240  is greater than  -922\n",
      "709  is greater than  403\n",
      "-112  is greater than  -113\n",
      "-381  is greater than  -555\n",
      "-462  is greater than  -617\n",
      "427  is greater than  -11\n",
      "424  is greater than  -469\n",
      "13  is greater than  -791\n",
      "-218  is greater than  -551\n",
      "178  is greater than  -224\n",
      "378  is greater than  246\n",
      "248  is greater than  -913\n",
      "792  is greater than  -400\n",
      "554  is greater than  -213\n",
      "792  is greater than  61\n",
      "574  is greater than  169\n",
      "771  is greater than  636\n",
      "444  is greater than  63\n",
      "-518  is greater than  -909\n",
      "-291  is greater than  -541\n",
      "253  is greater than  -903\n",
      "192  is greater than  181\n",
      "943  is greater than  -56\n",
      "566  is greater than  -93\n",
      "731  is greater than  490\n",
      "150  is greater than  31\n",
      "-311  is greater than  -774\n",
      "-517  is greater than  -726\n",
      "-418  is greater than  -644\n",
      "504  is greater than  -939\n",
      "929  is greater than  700\n",
      "994  is greater than  -607\n",
      "121  is greater than  -931\n",
      "976  is greater than  -673\n",
      "709  is greater than  -756\n",
      "659  is greater than  298\n",
      "-664  is greater than  -770\n",
      "407  is greater than  211\n",
      "-374  is greater than  -826\n",
      "464  is greater than  -766\n",
      "-382  is greater than  -786\n",
      "283  is greater than  -346\n",
      "510  is greater than  -863\n",
      "808  is greater than  -806\n",
      "731  is greater than  621\n",
      "666  is greater than  -158\n",
      "358  is greater than  -189\n",
      "366  is greater than  -479\n",
      "946  is greater than  225\n",
      "50  is greater than  -824\n",
      "224  is greater than  -513\n",
      "-685  is greater than  -717\n",
      "605  is greater than  413\n",
      "459  is greater than  -454\n",
      "453  is greater than  167\n",
      "2  is greater than  -370\n",
      "494  is greater than  360\n",
      "-268  is greater than  -363\n",
      "37  is greater than  -436\n",
      "857  is greater than  383\n",
      "-121  is greater than  -443\n",
      "845  is greater than  695\n",
      "509  is greater than  -960\n",
      "705  is greater than  -497\n",
      "744  is greater than  332\n",
      "944  is greater than  -922\n",
      "564  is greater than  -646\n",
      "391  is greater than  -951\n",
      "980  is greater than  -982\n",
      "-852  is greater than  -989\n",
      "884  is greater than  -892\n",
      "685  is greater than  203\n",
      "-256  is greater than  -648\n",
      "121  is greater than  -539\n",
      "-337  is greater than  -368\n",
      "-42  is greater than  -417\n",
      "890  is greater than  -153\n",
      "-113  is greater than  -682\n",
      "-460  is greater than  -865\n",
      "557  is greater than  -492\n",
      "636  is greater than  -793\n",
      "-430  is greater than  -809\n",
      "590  is greater than  564\n",
      "-498  is greater than  -917\n",
      "256  is greater than  -552\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_greater(greater)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Now write a function that returns the largest element on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def greatest(arr):\n",
    "    greather=0\n",
    "    for num in arr:\n",
    "        if num>greather:\n",
    "            greather=num\n",
    "    return greather\n",
    "\n",
    "\n",
    "\n",
    "o \n",
    "\n",
    "def greatest(arr):\n",
    "    return max(arr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.052s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_greatest(greatest)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Write a function that sums all the elements on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def sum_all(arr):\n",
    "    start=0\n",
    "    for i in arr:\n",
    "        start+= i\n",
    "    return start\n",
    "list=[1,2,3,4]\n",
    "sum_all(list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.052s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_sum(sum_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Write another function that multiplies all the elements on a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mult_all(arr):\n",
    "    mult_all = 1\n",
    "    for i in arr:\n",
    "        mult_tot *= i\n",
    "    return mult_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_mult' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [11]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will test your function \u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mtest_mult\u001b[49m(mult_all)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_mult' is not defined"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_mult(mult_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Now combine those two ideas and write a function that receives a list and either \"+\" or \"*\" and outputs acordingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def oper_all(arr, oper):\n",
    "    sum_total = 0\n",
    "    mult_total = 1\n",
    "    if oper == \"+\":\n",
    "        for i in arr:\n",
    "            sum_total += i\n",
    "        return sum_total\n",
    "    elif oper == \"*\":\n",
    "        for i in arr:\n",
    "            mult_total *= i\n",
    "        return mult_total\n",
    "    else:\n",
    "        pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will test your function \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Write a function that returns the factorial of a number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def factorial(n):\n",
    "    product = 1\n",
    "    for i in range(n):\n",
    "        product = product * (i+1)\n",
    "    return product\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "....................................................................................................\n",
      "----------------------------------------------------------------------\n",
      "Ran 100 tests in 0.048s\n",
      "\n",
      "OK\n"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_factorial(factorial)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Write a function that takes a list and returns a list of the unique values.\n",
    "\n",
    "`NOTE: You cannot use set. 🤔`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "expected an indented block (913347782.py, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Input \u001b[0;32mIn [28]\u001b[0;36m\u001b[0m\n\u001b[0;31m    freq_dict = {}\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m expected an indented block\n"
     ]
    }
   ],
   "source": [
    "def unique(arr):\n",
    "    def unique(arr):\n",
    "    freq_dict = {}\n",
    "    unique_lst = []\n",
    "    for i in arr:\n",
    "        if i in freq_dict:\n",
    "            freq_dict[i] += 1\n",
    "        else:\n",
    "            freq_dict[i] = 1\n",
    "    for key in freq_dict:\n",
    "        unique_lst.append(key)\n",
    "    return unique_lst\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_unique' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [29]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will test your function \u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mtest_unique\u001b[49m(unique)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_unique' is not defined"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_unique(unique)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Write a function that returns the mode of a list, i.e.: the element that appears the most times.\n",
    "`NOTE: You should not use count... 🧐`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mode_counter(arr):\n",
    "    frequency_dict = {}\n",
    "    max_value = 0\n",
    "    \n",
    "    for i in arr:\n",
    "        if i in frequency_dict:\n",
    "            frequency_dict[i] += 1\n",
    "        else:\n",
    "            frequency_dict[i] = 1\n",
    "    max_key = max(frequency_dict, key = frequency_dict.get)\n",
    "            \n",
    "    return max_key\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_mode' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [20]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will test your function \u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mtest_mode\u001b[49m(mode_counter)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_mode' is not defined"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_mode(mode_counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 9. Write a function that calculates the standard deviation of a list.\n",
    "`NOTE: Do not use any libraries or already built functions. 😉`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statistics import stdev\n",
    "def st_dev(arr):\n",
    "    prome = 0\n",
    "    varianza= 0\n",
    "    \n",
    "    for i in arr:\n",
    "        prome += i\n",
    "        \n",
    "    for i in arr:\n",
    "        promedio = prome/len(arr)\n",
    "        potencia = (i-promedio)**2\n",
    "        varianza += potencia\n",
    "    var = varianza/(len(arr)-1)\n",
    "    \n",
    "    return (var)**(1/2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_stdev' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [21]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will test your function \u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mtest_stdev\u001b[49m(st_dev)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_stdev' is not defined"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_stdev(st_dev)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10. Write a function to check if a string is a pangram, i.e.: if it contains all the letters of the alphabet at least once. Mind that the strings may contain characters that are not letters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def pangram(string):\n",
    "    alphabet = \"abcdefghijklmnopqrstuvwxyz\"\n",
    "    str_ = string.lower()\n",
    "    for char in alphabet:\n",
    "        if char not in str_:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_pangram' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [23]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will test your function \u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mtest_pangram\u001b[49m(pangram)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_pangram' is not defined"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_pangram(pangram)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 11. Write a function that receives a string of comma separated words and returns a string of comma separated words sorted alphabetically.\n",
    "\n",
    "`NOTE: You may use sorted but not split and definitely no join! 🤪`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Contraer' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [24]\u001b[0m, in \u001b[0;36m<cell line: 21>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m     newword2\u001b[38;5;241m=\u001b[39mnewword[\u001b[38;5;241m1\u001b[39m:]\n\u001b[1;32m     20\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m newword2\n\u001b[0;32m---> 21\u001b[0m \u001b[43mContraer\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Contraer' is not defined"
     ]
    }
   ],
   "source": [
    "def sort_alpha(string):\n",
    "    string+=\",\"\n",
    "    newlist=[]\n",
    "    word=\"\"\n",
    "    newword=\"\"\n",
    "    for letter in string.lower():\n",
    "        if letter ==\",\":\n",
    "            \n",
    "            newlist.append(word)\n",
    "            word=\"\"\n",
    "        else:\n",
    "            word+=letter\n",
    "            \n",
    "    newlist.sort()\n",
    "    \n",
    "    for elem in newlist:\n",
    "        newword+=\",\"+elem\n",
    "    \n",
    "    newword2=newword[1:]\n",
    "    return newword2\n",
    "Contraer\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_alpha' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [25]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will test your function \u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mtest_alpha\u001b[49m(sort_alpha)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_alpha' is not defined"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_alpha(sort_alpha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 12. Write a function to check if a given password is strong (at least 8 characters, at least one lower case, at least one upper case, at least one number and at least one special character). It should output True if strong and False if not.\n",
    "`Valid special characters: # @ ! $ % & ( ) ^ * [ ] { }`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Contraer' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [26]\u001b[0m, in \u001b[0;36m<cell line: 26>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     22\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m     24\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m count_lower \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m count_upper \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m count_special \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m count_number \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(string) \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m8\u001b[39m\n\u001b[0;32m---> 26\u001b[0m \u001b[43mContraer\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Contraer' is not defined"
     ]
    }
   ],
   "source": [
    "def check_pass(string):\n",
    "    count_lower = 0\n",
    "    count_upper = 0\n",
    "    count_special = 0\n",
    "    count_number = 0\n",
    "    \n",
    "    lower = \"abcdefghijklmnopqrstuvwxyz\"\n",
    "    upper = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'\n",
    "    special = \"#@!$%&()^*[]{}\"\n",
    "    number = '1234567890'\n",
    "    \n",
    "    for i in string:\n",
    "        if i in lower:\n",
    "            count_lower += 1\n",
    "        elif i in upper:\n",
    "            count_upper += 1\n",
    "        elif i in special:\n",
    "            count_special += 1\n",
    "        elif i in number:\n",
    "            count_number += 1\n",
    "        else:\n",
    "            return False\n",
    "            \n",
    "    return count_lower >= 1 and count_upper >= 1 and count_special >= 1 and count_number >= 1 and len(string) >= 8\n",
    "        \n",
    "Contraer\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'test_pass' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [27]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will test your function \u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mtest_pass\u001b[49m(check_pass)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'test_pass' is not defined"
     ]
    }
   ],
   "source": [
    "# This will test your function \n",
    "test_pass(check_pass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
